### ISSUE TITLE
Python机器学习实战：从数据预处理到模型部署

### ISSUE THREAD

**Issue by curtisreso** [curtisreso/Y#1]
1. 项目准备与环境搭建
1.1 安装必要库
!pip install numpy pandas scikit-learn matplotlib seaborn xgboost flask joblib
1.2 导入所需模块
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
import joblib
import warnings
warnings.filterwarnings('ignore')

# 设置随机种子保证可重复性
np.random.seed(42)
2. 数据加载与探索性分析
2.1 加载数据集
# 加载信用卡欺诈检测数据集
url = "https://raw.githubusercontent.com/nsethi31/Kaggle-Data-Credit-Card-Fraud-Detection/master/creditcard.csv"
data = pd.read_csv(url)

print(f"数据集形状: {data.shape}")
print("\n前5行数据:")
print(data.head())
2.2 数据探索
# 基本统计信息
print("\n基本统计信息:")
print(data.describe())

# 检查缺失值
print("\n缺失值统计:")
print(data.isnull().sum())

# 类别分布
plt.figure(figsize=(8, 5))
sns.countplot(x='Class', data=data)
plt.title('类别分布')
plt.show()

# 特征相关性
plt.figure(figsize=(12, 10))
sns.heatmap(data.corr(), cmap='coolwarm', center=0)
plt.title('特征相关性矩阵')
plt.show()
3. 数据预处理
3.1 特征工程
# 创建时间特征
data['Hour'] = data['Time'].apply(lambda x: x // 3600 % 24)
data['Day'] = data['Time'].apply(lambda x: x // (3600 * 24) % 7)

# 删除原始时间列
data = data.drop(['Time'], axis=1)

# 处理类别不平衡问题
from imblearn.over_sampling import SMOTE

X = data.drop('Class', axis=1)
y = data['Class']

# 使用SMOTE过采样
smote = SMOTE(random_state=42)
X_res, y_res = smote.fit_resample(X, y)

print("\n过采样后类别分布:")
print(pd.Series(y_res).value_counts())
3.2 数据分割与标准化
# 分割数据集
X_train, X_test, y_train, y_test = train_test_split(
    X_res, y_res, test_size=0.2, random_state=42)

# 数值特征标准化
numeric_features = X_train.columns.tolist()
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features)])

# 应用预处理
X_train = preprocessor.fit_transform(X_train)
X_test = preprocessor.transform(X_test)

print(f"\n训练集形状: {X_train.shape}")
print(f"测试集形状: {X_test.shape}")
4. 模型训练与评估
4.1 随机森林模型
# 训练随机森林
rf_model = RandomForestClassifier(
    n_estimators=100,
    max_depth=10,
    min_samples_leaf=5,
    random_state=42)

rf_model.fit(X_train, y_train)

# 评估模型
y_pred_rf = rf_model.predict(X_test)

print("\n随机森林分类报告:")
print(classification_report(y_test, y_pred_rf))

# 混淆矩阵
plt.figure(figsize=(6, 4))
sns.heatmap(confusion_matrix(y_test, y_pred_rf), 
            annot=True, fmt='d', cmap='Blues')
plt.title('随机森林混淆矩阵')
plt.show()
4.2 XGBoost模型
# 训练XGBoost
xgb_model = XGBClassifier(
    n_estimators=150,
    max_depth=8,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42)

xgb_model.fit(X_train, y_train)

# 评估模型
y_pred_xgb = xgb_model.predict(X_test)

print("\nXGBoost分类报告:")
print(classification_report(y_test, y_pred_xgb))

# 特征重要性
plt.figure(figsize=(12, 8))
feat_importances = pd.Series(
    xgb_model.feature_importances_, 
    index=numeric_features)
feat_importances.nlargest(15).plot(kind='barh')
plt.title('XGBoost特征重要性')
plt.show()
5. 模型优化与调参
5.1 网格搜索调参
from sklearn.model_selection import GridSearchCV

# 定义参数网格
param_grid = {
    'n_estimators': [50, 100, 150],
    'max_depth': [5, 8, 10],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.7, 0.8, 0.9]
}

# 创建网格搜索
grid_search = GridSearchCV(
    estimator=XGBClassifier(random_state=42),
    param_grid=param_grid,
    cv=3,
    scoring='f1',
    n_jobs=-1,
    verbose=1)

# 执行网格搜索
grid_search.fit(X_train, y_train)

# 输出最佳参数
print("\n最佳参数组合:")
print(grid_search.best_params_)

# 使用最佳模型
best_xgb = grid_search.best_estimator_
y_pred_best = best_xgb.predict(X_test)

print("\n优化后分类报告:")
print(classification_report(y_test, y_pred_best))
6. 模型部署与API创建
6.1 保存模型
# 保存预处理器和模型
joblib.dump(preprocessor, 'preprocessor.joblib')
joblib.dump(best_xgb, 'credit_card_fraud_model.joblib')

print("模型保存成功!")
6.2 创建Flask API
# app.py
from flask import Flask, request, jsonify
import joblib
import numpy as np

app = Flask(__name__)

# 加载模型和预处理器
preprocessor = joblib.load('preprocessor.joblib')
model = joblib.load('credit_card_fraud_model.joblib')

@app.route('/predict', methods=['POST'])
def predict():
    try:
        # 获取输入数据
        input_data = request.json
        
        # 转换为DataFrame
        input_df = pd.DataFrame([input_data])
        
        # 预处理
        processed_data = preprocessor.transform(input_df)
        
        # 预测
        prediction = model.predict(processed_data)
        proba = model.predict_proba(processed_data)[0]
        
        # 返回结果
        result = {
            'prediction': int(prediction[0]),
            'probability': float(proba[1]),
            'status': 'success'
        }
        
        return jsonify(result)
    
    except Exception as e:
        return jsonify({'status': 'error', 'message': str(e)})

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=True)
6.3 测试API
# 示例请求数据
sample_data = {
    "V1": -1.359807,
    "V2": -0.072781,
    "V3": 2.536347,
    "V4": 1.378155,
    "V5": -0.338321,
    "V6": 0.462388,
    "V7": 0.239599,
    "V8": 0.098698,
    "V9": 0.363787,
    "V10": 0.090794,
    "V11": -0.551600,
    "V12": -0.617801,
    "V13": -0.991390,
    "V14": -0.311169,
    "V15": 1.468177,
    "V16": -0.470401,
    "V17": 0.207971,
    "V18": 0.025791,
    "V19": 0.403993,
    "V20": 0.251412,
qd7z.cn/yijia
qd7z.cn/yingchao
qd7z.cn/xijia
qd7z.cn/ouguan
qd7z.cn/nba
qd7z.cn/cba
qd7z.cn/rizhilian
qd7z.cn/hanklian
qd7z.cn/ouguanbei
qd7z.cn/wnba
qd7z.cn/wcba
qd7z.cn/zhongchao
qd7z.cn/yazhoubei
qd7z.cn/ouguolian
qd7z.cn/meizhiye
qd7z.cn/meizhilian
qd7z.cn/ouzhoubei
qd7z.cn/shijiebei
qd7z.cn/yingzuzongbei
qd7z.cn/yaguan
qd7z.cn/yaguanbei
qd7z.cn/yaguanjingying
qd7z.cn/nbl
qd7z.cn/aojia
